engine/consensus/dkg/dkg_controller_test.go

```diff
@@ -0,0 +1,255 @@
+	id            int
+	controller    *Controller
+	phaseDuration time.Duration
+}
+
+func newNode(id int, controller *Controller, phaseDuration time.Duration) *node {
+	return &node{
+		id:            id,
+		controller:    controller,
+		phaseDuration: phaseDuration,
+	}
+}
+
+func (n *node) run() error {
+	// start the DKG controller
+	go func() {
+		_ = n.controller.Run()
+	}()
+
+	// Wait and trigger transition from Phase 0 to Phase 1
+	err := n.delayedTask(n.phaseDuration, n.controller.EndPhase0)
+	if err != nil {
+		return err
+	}
+
+	// Wait and trigger transition from Phase 1 to Phase 2
+	err = n.delayedTask(n.phaseDuration, n.controller.EndPhase1)
+	if err != nil {
+		return err
+	}
+
+	// Wait and retrieve DKG results
+	err = n.delayedTask(
+		n.phaseDuration,
+		func() error {
+			err := n.controller.End()
+			return err
+		})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (n *node) delayedTask(delay time.Duration, task func() error) error {
+	timer := time.After(delay)
+	<-timer
+	return task()
+}
+
+// processor is an implementation of DKGProcessor that enables nodes to exchange
+// private and public messages.
+type processor struct {
+	id       int
+	channels []chan DKGMessage
+}
+
+func (proc *processor) PrivateSend(dest int, data []byte) {
+	proc.channels[dest] <- DKGMessage{Orig: proc.id, Data: data}
+}
+
+func (proc *processor) Broadcast(data []byte) {
+	for i := 0; i < len(proc.channels); i++ {
+		if i == proc.id {
+			continue
+		}
+		proc.channels[i] <- DKGMessage{Orig: proc.id, Data: data}
+	}
+}
+
+func (proc *processor) Blacklist(node int) {}
+
+func (proc *processor) FlagMisbehavior(node int, logData string) {}
+
+type testCase struct {
+	totalNodes    int
+	phaseDuration time.Duration
+}
+
+// TestDKGNormal tests the controller in optimal conditions, when all nodes are
+// working correctly.
+func TestDKGNormal(t *testing.T) {
+	// define different test cases with varying number of nodes, and phase
+	// durations
+	testCases := []testCase{
+		testCase{totalNodes: 5, phaseDuration: time.Second},
+		testCase{totalNodes: 10, phaseDuration: time.Second},
+		testCase{totalNodes: 15, phaseDuration: 3 * time.Second},
+		testCase{totalNodes: 20, phaseDuration: 3 * time.Second},
+	}
+
+	// run each test case
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("%d nodes", tc.totalNodes), func(t *testing.T) {
+			testDKG(t, tc.totalNodes, tc.totalNodes, tc.phaseDuration)
+		})
+	}
+}
+
+// TestDKGThreshold tests that the controller results in a successful DKG as
+// long as the minimum threshold for non-byzantine nodes is satisfied.
+func TestDKGThreshold(t *testing.T) {
+	// define different test cases with varying number of nodes, and phase
+	// durations
+	testCases := []testCase{
+		testCase{totalNodes: 5, phaseDuration: time.Second},
+		testCase{totalNodes: 10, phaseDuration: time.Second},
+		testCase{totalNodes: 15, phaseDuration: 3 * time.Second},
+		testCase{totalNodes: 20, phaseDuration: 3 * time.Second},
```

Comment:
Since these are all happy path cases, there are no messages sent during phase 1 and 2, all messaging is done in phase 0. So we could have faster tests if phase 1 and 2 take like a short time.